File: packages/backend/cors.go
package main

import (
	"net/http"
	"os"
	"strings"

	"github.com/gin-gonic/gin"
)

// corsMiddleware sets CORS headers for local dev. Configure a specific
// origin via CORS_ALLOWED_ORIGINS (comma-separated list, or "*"). When
// credentials are used (the cookie-based auth in PRD-02) the response
// origin MUST be a single explicit value, never "*", so we echo the
// request origin whenever one is present.
func corsMiddleware() gin.HandlerFunc {
	allowed := strings.TrimSpace(os.Getenv("CORS_ALLOWED_ORIGINS"))
	allowedList := splitAndTrim(allowed)
	return func(c *gin.Context) {
		origin := c.Request.Header.Get("Origin")
		switch {
		case len(allowedList) == 0:
			// No allow-list: still echo a single origin (or "*") to keep credentials valid.
			if origin != "" {
				c.Header("Access-Control-Allow-Origin", origin)
				c.Header("Access-Control-Allow-Credentials", "true")
				c.Header("Vary", "Origin")
			} else {
				c.Header("Access-Control-Allow-Origin", "*")
			}
		case allowed == "*":
			if origin != "" {
				c.Header("Access-Control-Allow-Origin", origin)
				c.Header("Access-Control-Allow-Credentials", "true")
				c.Header("Vary", "Origin")
			} else {
				c.Header("Access-Control-Allow-Origin", "*")
			}
		default:
			if contains(allowedList, origin) && origin != "" {
				c.Header("Access-Control-Allow-Origin", origin)
				c.Header("Access-Control-Allow-Credentials", "true")
				c.Header("Vary", "Origin")
			}
		}
		c.Header("Access-Control-Allow-Methods", "GET, POST, PATCH, PUT, DELETE, OPTIONS")
		c.Header("Access-Control-Allow-Headers", "Content-Type, Authorization")
		c.Header("Access-Control-Max-Age", "86400")

		if c.Request.Method == http.MethodOptions {
			c.AbortWithStatus(http.StatusNoContent)
			return
		}
		c.Next()
	}
}

func splitAndTrim(s string) []string {
	if s == "" {
		return nil
	}
	parts := strings.Split(s, ",")
	out := make([]string, 0, len(parts))
	for _, p := range parts {
		p = strings.TrimSpace(p)
		if p != "" {
			out = append(out, p)
		}
	}
	return out
}

func contains(list []string, s string) bool {
	for _, x := range list {
		if x == s {
			return true
		}
	}
	return false
}

File: packages/backend/src/middleware/cors.ts
import type { Context, Next } from "hono";
import { env } from "../env.ts";

// Parses CORS_ALLOWED_ORIGINS into either:
//   - { mode: "open" }      — no allow-list configured (echo any origin or "*")
//   - { mode: "wildcard" }  — explicitly "*"
//   - { mode: "list", list } — specific allow-list
type Allowed = { mode: "open" } | { mode: "wildcard" } | { mode: "list"; list: string[] };

function parseAllowedOrigins(raw: string): Allowed {
  const parts = raw
    .split(",")
    .map((s) => s.trim())
    .filter(Boolean);
  if (parts.length === 0) return { mode: "open" };
  if (parts.length === 1 && parts[0] === "*") return { mode: "wildcard" };
  return { mode: "list", list: parts };
}

// Mirrors the Go `corsMiddleware`: when credentials are used (the cookie-based
// auth), the response Access-Control-Allow-Origin MUST be a single explicit
// value — never "*" — so we echo the request Origin whenever one is present.
export async function corsMiddleware(c: Context, next: Next) {
  const allowed = parseAllowedOrigins(env.CORS_ALLOWED_ORIGINS);
  const origin = c.req.header("Origin") ?? "";

  if (allowed.mode === "open" || allowed.mode === "wildcard") {
    if (origin) {
      c.header("Access-Control-Allow-Origin", origin);
      c.header("Access-Control-Allow-Credentials", "true");
      c.header("Vary", "Origin");
    } else {
      c.header("Access-Control-Allow-Origin", "*");
    }
  } else {
    if (origin && allowed.list.includes(origin)) {
      c.header("Access-Control-Allow-Origin", origin);
      c.header("Access-Control-Allow-Credentials", "true");
      c.header("Vary", "Origin");
    }
    // else: no header set — browser will block the request.
  }

  c.header("Access-Control-Allow-Methods", "GET, POST, PATCH, PUT, DELETE, OPTIONS");
  c.header("Access-Control-Allow-Headers", "Content-Type, Authorization");
  c.header("Access-Control-Max-Age", "86400");

  if (c.req.method === "OPTIONS") {
    return c.body(null, 204);
  }

  await next();
}
